### PR TITLE
Added error correction pipeline simulations

### PR DESCRIPTION
**Pull request by amicciche**
Implemented both naive and short syndrome error correction pipelines for both CSS and non-CSS codes. I have many non-official testing functions for these, but didn’t know if I should have included them. The main way I’ve been testing them is by generating plots and then inspecting them visually. Regardless this feature might be nice to have in the library, as currently these function just return a tuple of an x and z logical error rate for a given code and physical error rate. It would be nice if the user didn’t have to convert this into something more easy to plot themselves. Below I’ll include a sample function of mine, and based on discussion, might supply a second commit that adds these types of “nice-to-have” features.

Here is a sample testing function. There are some functions being called from my private repo, but hopefully the main idea of how to use this is clear.  This function in particular returns a plot for the logical X error, as well as a plot for the logical Z error.  (f_x and f_z are plots) 
```julia
function test_CSS_shor_refactor()
    name = "LDPC1"
    error_rates = exp10.(range(-5,-1.5,length=40))
    stab, Cx, Cz = CircuitCompilation2xn.getGoodLDPC(1)
    code = QuantumClifford.ECC.CSS(stab, Cx, Cz)
    post_ec_error_rates = [QuantumClifford.ECC.CSS_shor_error_correction_pipeline(code, p, nframes=10_000) for p in error_rates]

    x_error = [post_ec_error_rates[i][1] for i in eachindex(post_ec_error_rates)]
    z_error = [post_ec_error_rates[i][2] for i in eachindex(post_ec_error_rates)]

    f_x = CircuitCompilation2xn.plot_code_performance_log_log(error_rates, x_error,title="Logical X Error of "*name*" Circuit Krishna")
    f_z = CircuitCompilation2xn.plot_code_performance_log_log(error_rates, z_error,title="Logical Z Error of "*name*" Circuit Krishna")
    return f_x, f_z
end
```

And its output:
image
image

### CODE REVIEW

**Comment by Krastanov** [comment]
A lot of the failures are due to the LDPCDecoders dependency. I cleaned it up a bit. Let's see whether the tests work better after https://github.com/JuliaRegistries/General/pull/97229

**Comment by Krastanov** [comment]
@Benzillaist, could you drop a link here to the code you have prepared for the wiki. If I remember correctly, we had discussed some API design with you, and it would be useful for me to refer to that API while working through reviewing this code.

**Comment by Krastanov** [comment]
There were some changes done on the master branch which help with testing. I updated this branch to include the changes from master. Be sure to first do `git pull` before making further changes to this branch, to avoid divergences.

File: src/ecc/pipelines.jl [hunk 1]
```diff
@@ -0,0 +1,512 @@
+"""Generate a lookup table for decoding single qubit errors."""
+function create_lookup_table(code::Stabilizer)
```
Comment (Krastanov) [inline src/ecc/pipelines.jl:2]: @Benzillaist could you link to the implementation you have for the lookup table? In particular in the type-based API for decoders

**Comment by Krastanov** [comment]
For these four functions we need a better API:

```
naive_error_correction_pipeline
shor_error_correction_pipeline
CSS_naive_error_correction_pipeline
CSS_shor_error_correction_pipeline
```

One type of code repetition here is that there are two copies of the simulation code for each of the hardcoded decoders. It would be useful to decouple the decoders.

Something that Ben (@Benzillaist) has been working on is a decoder API.

@Ben, could you submit a PR to LDPCDecoders with the API?

It is along the lines of:

```
decoder_instance_x = SomeDecoderType(configuration...)
decoder_instance_z = SomeDecoderType(configuration...)
predicted_error_pattern_x = decode(decoder_instance_x, syndrome_measurement_x)
predicted_error_pattern_z = decode(decoder_instance_z, syndrome_measurement_z)
predicted_error_pattern = cat(...)
decoded += logical_measurements == faults_matrix[??] * predicted_error_pattern
```
(done once for the X logicals and once for the Z logicals)

**Comment by Benzillaist** [comment]
Hello Stefan, apologies for the delay. I have made a pull request to LDPCDecoders with my code as it is right now.
I have unfortunately had issues with calling the faults_matrix method on the Stabilizers that I create, meaning that I am finding it very hard to test my code. I would greatly appreciate someone looking at my code to help me debug where I am doing something wrong with regards to getting the faults matrix.

Link to PR: https://github.com/QuantumSavory/LDPCDecoders.jl/pull/11

**Comment by Krastanov** [comment]
superseded by #212 -- much of #212 is based on what was here

**Pull request closed without merging**